### ISSUE TITLE
Debugability of QuestPDF.Drawing.Exceptions.DocumentDrawingException

### ISSUE THREAD

**Issue by luetm** [QuestPDF/QuestPDF#308]
**Is your feature request related to a problem? Please describe.**
General debugability of quest pdf.

**Describe the solution you'd like**
If possible, don't throw at GeneratePdf() but in the lambdas themselves, or at least give better pointers where the problem could be. In complex cases it's a needle in a haystack.

**Describe alternatives you've considered**
n/a

**Additional context**
How would I go about debugging something like this?

```
QuestPDF.Drawing.Exceptions.DocumentDrawingException
Inner: Value cannot be null. (Parameter 'text')

Outer stacktrace: 
   at QuestPDF.Drawing.DocumentGenerator.RenderPass[TCanvas](PageContext pageContext, TCanvas canvas, Container content, DocumentMetadata documentMetadata, DebuggingState debuggingState)
   at QuestPDF.Drawing.DocumentGenerator.RenderDocument[TCanvas](TCanvas canvas, IDocument document)
   at QuestPDF.Drawing.DocumentGenerator.GeneratePdf(Stream stream, IDocument document)
   at QuestPDF.Fluent.GenerateExtensions.GeneratePdf(IDocument document, Stream stream)
   at QuestPDF.Fluent.GenerateExtensions.GeneratePdf(IDocument document, String filePath)
   at Program.<<Main>$>d__0.MoveNext() in C:\Dev\FullKgTest\FullKgTest\Program.cs:line 37

Inner stacktrace:
   at SkiaSharp.SKCanvas.DrawText(SKTextBlob text, Single x, Single y, SKPaint paint)
   at QuestPDF.Drawing.SkiaCanvasBase.DrawText(SKTextBlob skTextBlob, Position position, TextStyle style)
   at QuestPDF.Elements.Text.Items.TextBlockSpan.Draw(TextDrawingRequest request)
   at QuestPDF.Elements.Text.TextBlock.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Column.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Padding.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Background.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Table.Table.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Decoration.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Background.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Column.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Column.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Decoration.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Padding.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Constrained.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Layers.<>c__DisplayClass6_0.<Draw>b__1(Layer x)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at QuestPDF.Elements.Layers.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Background.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Elements.Column.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Infrastructure.ContainerElement.Draw(Size availableSpace)
   at QuestPDF.Drawing.DocumentGenerator.RenderPass[TCanvas](PageContext pageContext, TCanvas canvas, Container content, DocumentMetadata documentMetadata, DebuggingState debuggingState)

```

From the inner stacktrace I take home that the problem is probably in a .Span(), however I checked them all for null and they all have a ?? " " now but I still get this error. I have no idea what is going wrong. So now I can look forward to spending hours commenting my entire code and then uncomment line by line to see where the problem is. This is the only, but a big gripe I have with this library (which otherwise is amazing, thank you!).

**Comment by MarcinZiabek** [comment]
What version of QuestPDF are you using? Would you be able to locate the exact place of this exception? My assumption is that the `null` value should be treated exactly as an empty string.

**Comment by luetm** [comment]
I'm using 2022.08.0. I had to increase the width of a ConstantColumn in a table

```
container.Column(column =>
{
    column.Item().Text(Strings.Files ?? " ").Style(H3);
    column.Spacing(DefaultSpacing);
    column.Item().DefaultTextStyle(ContentSmall).Background(BackgroundTable1).Table(table =>
    {
        table.ColumnsDefinition(c =>
        {
            c.ConstantColumn(60);
            c.RelativeColumn();
            c.ConstantColumn(150);
            c.ConstantColumn(200);
        });

        table.Header(header =>
        {
            foreach (var title in TitleRow)
            {
                header.Cell().Background(BackgroundBoxDarker).Padding(4).Text(title ?? " ").SemiBold();
            }
        });

        int i = 0;
        foreach (var row in Rows)
        {
            var background = i++ % 2 == 0 ? BackgroundTable1 : BackgroundTable2;
            table.Cell().Background(background).PaddingVertical(3).PaddingLeft(4).Text(row.Date ?? " ");
            table.Cell().Background(background).PaddingVertical(3).PaddingLeft(4).Text(row.Title ?? " ");
            table.Cell().Background(background).PaddingVertical(3).PaddingLeft(4).Text(row.Category ?? " ");
            table.Cell().Background(background).PaddingVertical(3).PaddingLeft(4).Text(row.FileName ?? " ");
        }
    });
});
```

I increased the column from 180 to 200 and now it works. Probably this is the reason for the other error I'm getting also in another table. Why does this happen?

**Comment by Ducki** [comment]
Unfortunately, I need to second this, especially when getting a DocumentLayoutException. There is literally nothing I can do to narrow it down to where something breaks. For some reason, I never get the detailed element trace in the console as documented.

**Comment by luetm** [comment]
For me, I got a lot of them when doing a PageBreak. When I have a few items, and all items have dynamic content but I put one page per item, when the item would just use up one page exactly, when doing a page break I get a DocumentLayoutException.

**Comment by Ducki** [comment]
Ooohhh, that was the important missing part: "only with debugger attached". I'll make a PR shortly to get this in the docs :-)

I was able to narrow it down with this (looks like DefaultTextStyle() handling has changed in the 2022.8 release, that reset some font sizes in my case). Thank you!

**Comment by luetm** [comment]
@MarcinZiabek 
```
                                                    🔥 DebuggingProxy
                                                    -----------------
                                                    Available space: (Width: 535.400, Height: -2.534)
                                                    Requested space: PartialRender (Width: 0.000, Height: 0.000)

                                                        🔥 PageBreak
                                                        ------------
                                                        Available space: (Width: 535.400, Height: -2.534)
                                                        Requested space: PartialRender (Width: 0.000, Height: 0.000)
```

This is what the last line looks like. It appears, that the Page break doesn't have a height left. How do I handle this case?

**Comment by MarcinZiabek** [comment]
@luetm Would you like to post entire ElementTrace? E.g. by using GitHub gist?

**Comment by luetm** [comment]
https://gist.github.com/luetm/155e39e0cbb985585ab5ef8f56787ed8

There is goes.

**Comment by MarcinZiabek** [comment]
@luetm I feel that this may be a problem with floating number arithmetic. When comparing two floats, I am using `Epsilon = 0.001pt`. This is the maximum absolute difference used for equality. 

I am able to force the library to throw the DocumentLayoutException with my sample report, when I change the `Epsilon = 0.00001pt`, so significantly smaller. However, the same effect could be observed with current Epsilon, when having a lot of content on the page with deep structures. This gives plenty of opportunities to accumulute rounding errors. 

I can change `Epsilon = 0.01f`. This still should be plenty but is it enough? I don't want to use doubles because of the memory allocations. Numbers are everywhere in QuestPDF.

What do you think?

**Comment by luetm** [comment]
I'm happy to test that! If you point me to the places in the code where I can change the epsilon, I can build and see what happens.

**Comment by girlpunk** [comment]
> Available space: (Width: 535.400, Height: -2.534)

I never managed to track down what was causing it, but had similar issues in the past (#206). As a workaround, try explicitly adding 0pt of padding immediately before the page break.

**Comment by MarcinZiabek** [comment]
I am  pretty sure that I have found the root cause yesterday. After all, it is not connected to the Epsilon value. Rather, it is a very specific corner case in the `PageBreak` code that needs to handle negative available space, [take look at this commit](https://github.com/QuestPDF/QuestPDF/commit/f0ba5fc32d9ac0187a22bcf72e7627e33aeca067). 

I am hoping to publish the next release soon 😁

**Comment by MarcinZiabek** [comment]
These bugs should be fixed in the 2022.8.1 release 😁

**Comment by luetm** [comment]
@MarcinZiabek Perfect! Now the batch runs for all 2.7k documents without error! (tested with 2022.8.2)
